cni/network/network.go

```diff
@@ -1008,83 +1046,89 @@ func (plugin *NetPlugin) Delete(args *cniSkel.CmdArgs) error {
+		// otherwise, in stateless, we don't need the network id for deletion
+		epInfos, err = plugin.nm.GetEndpointState(networkID, args.ContainerID)
+	} else {
+		epInfos = plugin.nm.GetEndpointInfosFromContainerID(args.ContainerID)
+	}
+
+	// for when the endpoint is not created, but the ips are already allocated (only works if single network, single infra)
+	// stateless cni won't have this issue
+	if len(epInfos) == 0 {
 		endpointID := plugin.nm.GetEndpointID(args.ContainerID, args.IfName)
-		// Query the endpoint.
-		if epInfo, err = plugin.nm.GetEndpointInfo(networkID, endpointID); err != nil {
-			logger.Info("GetEndpoint",
+		if !nwCfg.MultiTenancy {
+			logger.Error("Failed to query endpoint",
 				zap.String("endpoint", endpointID),
 				zap.Error(err))
-			if !nwCfg.MultiTenancy {
-				// attempt to release address associated with this Endpoint id
-				// This is to ensure clean up is done even in failure cases
-
-				logger.Error("Failed to query endpoint",
-					zap.String("endpoint", endpointID),
-					zap.Error(err))
+			if !plugin.nm.IsStatelessCNIMode() {
 				logger.Error("Release ip by ContainerID (endpoint not found)",
 					zap.String("containerID", args.ContainerID))
 				sendEvent(plugin, fmt.Sprintf("Release ip by ContainerID (endpoint not found):%v", args.ContainerID))
 				if err = plugin.ipamInvoker.Delete(nil, nwCfg, args, nwInfo.Options); err != nil {
 					return plugin.RetriableError(fmt.Errorf("failed to release address(no endpoint): %w", err))
 				}
 			}
-			// Log the error but return success if the endpoint being deleted is not found.
-			err = nil
-			return err
 		}
-
-		// schedule send metric before attempting delete
-		defer sendMetricFunc() //nolint:gocritic
-		logger.Info("Deleting endpoint",
-			zap.String("endpointID", endpointID))
-		sendEvent(plugin, fmt.Sprintf("Deleting endpoint:%v", endpointID))
-		// Delete the endpoint.
-		if err = plugin.nm.DeleteEndpoint(networkID, endpointID, epInfo); err != nil {
+		// Log the error but return success if the endpoint being deleted is not found.
+		err = nil
+		return err
+	}
+	logger.Info("Deleting the endpoints", zap.Any("endpointInfos", epInfos))
+	// populate ep infos here in loop if necessary
+	// delete endpoints
+	for _, epInfo := range epInfos {
+		// in stateless, network id is not populated in epInfo, but in stateful cni, it is (nw id is used in stateful)
+		if err = plugin.nm.DeleteEndpoint(epInfo.NetworkID, epInfo.EndpointID, epInfo); err != nil {
+			// An error will not be returned if the endpoint is not found
 			// return a retriable error so the container runtime will retry this DEL later
 			// the implementation of this function returns nil if the endpoint doens't exist, so
 			// we don't have to check that here
 			return plugin.RetriableError(fmt.Errorf("failed to delete endpoint: %w", err))
 		}
+	}
+	logger.Info("Deleting the endpoints from the ipam")
+	// delete endpoint state in cns and in statefile
+	for _, epInfo := range epInfos {
```

Comment:
why we add a separate loop instead call ipam delete in same loop?

Reply:
I believe ipam invoker delete deletes the state in stateless mode so if we need to perform, for example, 3 deletes and fail during the second DeleteEndpoint call, we may end up in an inconsistent state. I believe you requested that all endpoints be deleted first before modifying the state (whether it be stored on the cns like in stateless or in the cni statefile).